bootstrap/v2/pkg/apis/apps/kfdef/v1alpha1/application_types.go

```diff
@@ -179,3 +216,125 @@ func GetDefaultRegistry() *RegistryConfig {
+func DownloadAndLoadConfigFile(configFile string, appDir string) (*KfDef, error) {
+	if configFile == "" {
+		return nil, fmt.Errorf("config file must be the URI of a KfDef spec")
+	}
+
+	if appDir == "" {
+		return nil, fmt.Errorf("appdir must be the directory where the app should be stored")
+	}
+
+	if _, err := os.Stat(appDir); os.IsNotExist(err) {
+		log.Infof("Creating directory %v", appDir)
+		appdirErr := os.MkdirAll(appDir, os.ModePerm)
+		if appdirErr != nil {
+			log.Errorf("couldn't create directory %v Error %v", appDir, appdirErr)
+			return nil, appdirErr
+		}
+	} else {
+		log.Infof("App directory exists %v", appDir)
+	}
+
+
+	// Open config file
+	//
+	// TODO(jlewi): Should we use hashicorp go-getter.GetAny here? We use that to download
+	// the tarballs for the repos. Maybe we should use that here as well to be consistent.
+	appFile := path.Join(appDir, KfConfigFile)
+
+	if _, err:= os.Stat(appFile); err == nil {
+		log.Infof("%v exists not refetching", appFile)
+	} else {
+		log.Infof("Downloading %v to %v", configFile, appFile)
+		err := gogetter.GetFile(appFile, configFile)
+		if err != nil {
+			return nil, &kfapis.KfError{
+				Code:    int(kfapis.INVALID_ARGUMENT),
+				Message: fmt.Sprintf("could not fetch specified config %s: %v", configFile, err),
+			}
+		}
+	}
+	// Read contents
+	configFileBytes, err := ioutil.ReadFile(appFile)
+	if err != nil {
+		return nil, &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: fmt.Sprintf("could not read from config file %s: %v", configFile, err),
+		}
+	}
+	// Unmarshal content onto KfDef struct
+	kfDef := &KfDef{}
+	if err := yaml.Unmarshal(configFileBytes, kfDef); err != nil {
+		return nil, &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: fmt.Sprintf("could not unmarshal config file onto KfDef struct: %v", err),
+		}
+	}
+
+	// appDir should be stored in the resulting KfDef
+	kfDef.Spec.AppDir = appDir
+	return kfDef, nil
+}
+
+// SyncCache will synchronize the local cache of any repositories.
+// On success the status is updated with pointers to the cache.
+func (d *KfDef) SyncCache() error {
+	if d.Spec.AppDir == "" {
+		return fmt.Errorf("AppDir must be specified")
+	}
+
+	if d.Status.ReposCache == nil {
+		d.Status.ReposCache = make(map[string]RepoCache)
+	}
+	appDir := d.Spec.AppDir
+	// Loop over all the repos and download them.
+	// TODO(jlewi): We should check if we already have a local copy and not redownload it.
+
+	baseCacheDir := path.Join(appDir, DefaultCacheDir)
+	if _, err := os.Stat(baseCacheDir); os.IsNotExist(err) {
+		log.Infof("Creating directory %v", baseCacheDir)
+		appdirErr := os.MkdirAll(baseCacheDir, os.ModePerm)
+		if appdirErr != nil {
+			log.Errorf("couldn't create directory %v Error %v", baseCacheDir, appdirErr)
+			return appdirErr
+		}
+	}
+
+	for _, r := range d.Spec.Repos {
+		cacheDir := path.Join(baseCacheDir, r.Name)
+		// idempotency
+		// TODO(jlewi): Why do we remove the directory if it exists?
```

Comment:
Can we track commit hash of existing repos to avoid downloading same repo?

Reply:
Yes we should do something intelligent to avoid redownloading the repos.
Filed #3545